File: packages/contracts/src/notification.ts
/**
 * Notification types shared between the mobile app and the Go API.
 *
 * These mirror `internal/notifications` and docs/08-notifications-and-background.md.
 * The vocabulary is singular: the server's `ReminderTypes` and the union below
 * must agree, and a test in the API asserts the server half of that
 * (plans/phase8.md §2).
 */

/**
 * The kinds of reminder MeraCare schedules.
 *
 * Exactly the three docs/08 assigns to device-local scheduled reminders. An
 * event that needs telling somebody about right now — a missed dose, an
 * invitation — is a push notification, which is different infrastructure and a
 * later phase.
 */
export const REMINDER_TYPES = [
  'TASK_REMINDER',
  'MEDICATION_REMINDER',
  'APPOINTMENT_REMINDER',
] as const;
export type ReminderType = (typeof REMINDER_TYPES)[number];

/** What a reminder points at, so the app can open the right screen. */
export const REMINDER_ENTITIES = ['task_instance', 'medication_dose', 'appointment'] as const;
export type ReminderEntity = (typeof REMINDER_ENTITIES)[number];

/**
 * One notification the device should schedule locally.
 *
 * Deliberately carries no title and no body. Both are composed on the device
 * from `notification-labels.ts`, which is what keeps a medicine's name out of
 * anything that can appear on a lock screen: the server never sends it, so no
 * screen can accidentally show it (plans/phase8.md §§16, 17, 47).
 */
export interface Reminder {
  /**
   * Stable for as long as the reminder means the same thing, and different the
   * moment it does not. Used directly as the OS notification identifier, which
   * is what makes rescheduling idempotent (plans/phase8.md §25).
   */
  id: string;
  type: ReminderType;

  seniorId: string;
  seniorName: string;
  /** IANA name. The senior's clock decides what "08:00" means, not the reader's. */
  seniorTimezone: string;

  entityType: ReminderEntity;
  entityId: string;

  /** When the care itself is due, ISO-8601. */
  dueAt: string;
  /** When the notification should appear, ISO-8601. */
  fireAt: string;
}

/** A complete reminder plan: everything this user's device should have scheduled. */
export interface ReminderPlan {
  reminders: Reminder[];
  generatedAt: string;
  /**
   * Nothing beyond this instant is planned, so a scheduled notification firing
   * later than this must not be cancelled merely for being absent from the plan.
   */
  horizonEndsAt: string;
}

/**
 * Which categories of notification a user wants.
 *
 * Per user, not per senior: two caregivers looking after the same person
 * routinely want different things (plans/phase8.md §§3, 4).
 */
export interface NotificationPreferences {
  taskReminders: boolean;
  medicationReminders: boolean;
  appointmentReminders: boolean;
  /** Phase 11: a task whose time has passed with nothing recorded. */
  overdueTaskAlerts: boolean;
  /** Phase 11: something somebody else did. */
  careActivity: boolean;
  updatedAt: string;
}

/** One category the settings screen can switch. */
export interface NotificationCategory {
  key: keyof Omit<NotificationPreferences, 'updatedAt'>;
  label: string;
  description: string;
}

/**
 * The switches on the notification settings screen, in order.
 *
 * Worded as what the user gets, not as what the system does: nobody is choosing
 * a `MEDICATION_REMINDER`, they are choosing whether their phone tells them
 * about medicine (plans/phase8.md §18).
 */
export const NOTIFICATION_CATEGORIES: NotificationCategory[] = [
  {
    key: 'medicationReminders',
    label: 'Medication reminders',
    description: 'A reminder shortly before each dose is due.',
  },
  {
    key: 'taskReminders',
    label: 'Care task reminders',
    description: 'A reminder shortly before a care task is due.',
  },
  {
    key: 'appointmentReminders',
    label: 'Appointment reminders',
    description: 'A reminder the day before, and an hour before it starts.',
  },
  {
    key: 'overdueTaskAlerts',
    label: 'Overdue task alerts',
    description: 'A nudge when a care task has passed its time with nothing recorded.',
  },
  {
    key: 'careActivity',
    label: 'Care activity',
    description: 'When someone else records care — a dose given, a task completed.',
  },
];

/** The operating systems a device registration can name. */
export const DEVICE_PLATFORMS = ['ios', 'android', 'web'] as const;
export type DevicePlatform = (typeof DEVICE_PLATFORMS)[number];

/** What the app sends when it announces an installation. */
export interface DeviceRegistration {
  deviceId: string;
  platform: DevicePlatform;
  /** Absent until the user has granted OS notification permission. */
  pushToken?: string;
  appVersion?: string;
}

/**
 * A registered installation, as the API describes it.
 *
 * There is no push token field, and adding one would be a security regression:
 * the client that registered the token already has it, and no other client has
 * any business reading it (plans/phase8.md §8).
 */
export interface RegisteredDevice {
  id: string;
  deviceId: string;
  platform: DevicePlatform;
  appVersion: string;
  active: boolean;
  lastSeenAt: string;
  pushTokenRegistered: boolean;
}

/**
 * The kinds of notification MeraCare delivers and keeps.
 *
 * A superset of `REMINDER_TYPES`: the three a device can schedule for itself,
 * plus the two only a server can know about. Mirrors
 * `internal/notifications.Types` and the CHECK constraint on
 * `notifications.notification_type` (plans/phase11.md §5).
 */
export const NOTIFICATION_TYPES = [
  'MEDICATION_REMINDER',
  'APPOINTMENT_REMINDER',
  'TASK_REMINDER',
  'TASK_OVERDUE',
  'CARE_ACTIVITY',
] as const;
export type NotificationType = (typeof NOTIFICATION_TYPES)[number];

/** What a notification points at, so the app can open the right screen. */
export const NOTIFICATION_ENTITIES = [
  'task_instance',
  'medication_dose',
  'appointment',
  'care_event',
] as const;
export type NotificationEntity = (typeof NOTIFICATION_ENTITIES)[number];

/**
 * One row of the notification inbox.
 *
 * Unlike a `Reminder`, this one *does* carry its words — because they are the
 * words that were sent. A reminder is a plan the device renders; a notification
 * is a record of something that happened, and re-composing its sentence later
 * would make yesterday's inbox change when an appointment moves
 * (plans/phase11.md §6).
 *
 * The wording still says nothing medical: the same lock-screen privacy policy
 * produced it, on the server, in `internal/notifications/wording.go`
 * (plans/phase11.md §48).
 */
export interface Notification {
  id: string;
  type: NotificationType;

  title: string;
  body: string;

  seniorId: string;
  entityType: NotificationEntity;
  entityId: string;

  /** What the notification is for, ISO-8601. Also the inbox's sort order. */
  occurredAt: string;
  read: boolean;
  /** Empty while unread. */
  readAt: string;
}

/** One page of the inbox, newest first. */
export interface NotificationInbox {
  items: Notification[];
  nextCursor: string | null;
  /**
   * Every unread notification, not just the ones on this page — it travels with
   * the page so a badge and the list it labels cannot disagree
   * (plans/phase11.md §61).
   */
  unreadCount: number;
}

/** What marking everything read changed. */
export interface MarkAllReadResult {
  markedRead: number;
  unreadCount: number;
}

/**
 * The identifiers a push notification carries.
 *
 * Identifiers only, and never treated as permission. Tapping a notification
 * opens a screen, and that screen asks the server for the thing under the
 * reader's own authorization — so a notification that outlived somebody's
 * access opens nothing (plans/phase11.md §§31, 58).
 */
export interface PushPayload {
  notificationId: string;
  type: NotificationType;
  seniorId: string;
  entityType: NotificationEntity;
  entityId: string;
}

/**
 * Reads a payload off a push notification the user tapped.
 *
 * Returns null rather than throwing for anything unrecognised, for the same
 * reason `readReminderPayload` does: a notification can outlive the app version
 * that sent it, and a crash on opening a stale one is a worse bug than a tap
 * that does nothing.
 */
export function readPushPayload(data: unknown): PushPayload | null {
  if (typeof data !== 'object' || data === null) return null;

  const candidate = data as Record<string, unknown>;
  const required = ['notificationId', 'type', 'seniorId', 'entityType', 'entityId'] as const;

  for (const key of required) {
    if (typeof candidate[key] !== 'string' || candidate[key] === '') return null;
  }

  if (!NOTIFICATION_TYPES.includes(candidate.type as NotificationType)) return null;
  if (!NOTIFICATION_ENTITIES.includes(candidate.entityType as NotificationEntity)) return null;

  return {
    notificationId: candidate.notificationId as string,
    type: candidate.type as NotificationType,
    seniorId: candidate.seniorId as string,
    entityType: candidate.entityType as NotificationEntity,
    entityId: candidate.entityId as string,
  };
}

File: apps/api/internal/notifications/notification.go
// Package notifications decides what MeraCare is allowed to remind each user
// about, and when.
//
// It is infrastructure, not care. Tasks, medication, and appointments remain
// the only source of truth for what is due; this package reads their schedules
// and answers a narrower question — "which of those should this particular
// person's phone tell them about, and at what moment?" It creates no care data
// of its own and can be deleted without losing any (plans/phase8.md §1).
package notifications

import (
	"errors"
	"fmt"
	"slices"
	"time"

	"github.com/google/uuid"
)

// ReminderType is the kind of thing a reminder is about.
//
// These are the three categories docs/08-notifications-and-background.md
// assigns to scheduled local reminders, and the three plans/phase8.md §2
// requires. There is deliberately no fourth: an event that needs telling
// somebody about right now is a push notification, which is a different
// mechanism and a later phase.
type ReminderType string

const (
	// ReminderTaskReminder is a care task falling due.
	ReminderTaskReminder ReminderType = "TASK_REMINDER"
	// ReminderMedicationReminder is a dose falling due.
	ReminderMedicationReminder ReminderType = "MEDICATION_REMINDER"
	// ReminderAppointmentReminder is an appointment starting soon.
	ReminderAppointmentReminder ReminderType = "APPOINTMENT_REMINDER"
)

// ReminderTypes lists every reminder type. The mobile client's vocabulary in
// packages/contracts/src/notification.ts must match this exactly.
var ReminderTypes = []ReminderType{
	ReminderTaskReminder,
	ReminderMedicationReminder,
	ReminderAppointmentReminder,
}

// Valid reports whether the type is recognised.
func (t ReminderType) Valid() bool { return slices.Contains(ReminderTypes, t) }

// leadTime is how long before the thing itself a reminder fires.
//
// Fixed, and not configurable. Nothing in the documentation defines reminder
// offsets, and plans/phase8.md §12 forbids inventing options beyond the
// documented MVP — so these are exactly the three worked examples the brief
// itself gives: a dose at 08:00 reminds at 07:45 (§12), a task due 09:00
// reminds at 08:45 (§13), an appointment at 14:00 reminds at 13:00 (§14).
//
// The asymmetry is real rather than an oversight. Fifteen minutes is enough
// warning to walk to the kitchen; an appointment needs enough warning to leave
// the house.
func (t ReminderType) leadTime() time.Duration {
	switch t {
	case ReminderAppointmentReminder:
		return time.Hour
	default:
		return 15 * time.Minute
	}
}

// EntityType is what a reminder points at, so the app can open the right
// screen. The names match the domains' own vocabulary.
type EntityType string

const (
	// EntityTaskInstance is one occurrence of a care task.
	EntityTaskInstance EntityType = "task_instance"
	// EntityMedicationDose is one scheduled dose.
	EntityMedicationDose EntityType = "medication_dose"
	// EntityAppointment is one appointment.
	EntityAppointment EntityType = "appointment"
)

// entityFor is the subject each reminder type is about. One reminder type has
// exactly one entity type, so the client never has to handle a combination
// that cannot occur.
func (t ReminderType) entityFor() EntityType {
	switch t {
	case ReminderTaskReminder:
		return EntityTaskInstance
	case ReminderMedicationReminder:
		return EntityMedicationDose
	default:
		return EntityAppointment
	}
}

// Platform is the operating system a registered device runs.
type Platform string

const (
	// PlatformIOS is an iPhone or iPad.
	PlatformIOS Platform = "ios"
	// PlatformAndroid is an Android phone or tablet.
	PlatformAndroid Platform = "android"
	// PlatformWeb is the browser build. It registers like any other device;
	// whether a browser can be pushed to is the push phase's problem.
	PlatformWeb Platform = "web"
)

// Platforms lists every recognised platform. The database CHECK on
// notification_devices.platform must match.
var Platforms = []Platform{PlatformIOS, PlatformAndroid, PlatformWeb}

// Valid reports whether the platform is recognised.
func (p Platform) Valid() bool { return slices.Contains(Platforms, p) }

// Preferences are one user's notification settings.
//
// Held per user rather than per senior: the person receiving the notification
// is the person who decides whether they want it (plans/phase8.md §§3, 4). Two
// caregivers in the same circle can and routinely will differ.
type Preferences struct {
	UserID uuid.UUID

	TaskReminders        bool
	MedicationReminders  bool
	AppointmentReminders bool
	// OverdueTaskAlerts and CareActivity are Phase 11's two additions. 0008
	// refused to store switches for categories nothing could deliver; these two
	// now have a delivery path, so they exist (plans/phase11.md §9).
	OverdueTaskAlerts bool
	CareActivity      bool

	CreatedAt time.Time
	UpdatedAt time.Time
}

// DefaultPreferences is what a user has before they have ever opened the
// settings screen. Every category is on: reminders are the reason the app
// exists, and silence by default would look like a broken app rather than a
// respectful one.
func DefaultPreferences(userID uuid.UUID) Preferences {
	return Preferences{
		UserID:               userID,
		TaskReminders:        true,
		MedicationReminders:  true,
		AppointmentReminders: true,
		OverdueTaskAlerts:    true,
		CareActivity:         true,
	}
}

// wants reports whether these preferences allow the given reminder type.
func (p Preferences) wants(t ReminderType) bool {
	return p.wantsType(t.notificationType())
}

// wantsType reports whether these preferences allow the given notification
// type.
//
// The default is false, not true. A category nobody has decided about is a
// category MeraCare has not been given permission to use, and an unrecognised
// type reaching here at all means the vocabulary has grown without the
// preferences following it.
func (p Preferences) wantsType(t Type) bool {
	switch t {
	case TypeTaskReminder:
		return p.TaskReminders
	case TypeMedicationReminder:
		return p.MedicationReminders
	case TypeAppointmentReminder:
		return p.AppointmentReminders
	case TypeTaskOverdue:
		return p.OverdueTaskAlerts
	case TypeCareActivity:
		return p.CareActivity
	default:
		return false
	}
}

// Device is one installation of the app that may be pushed to.
type Device struct {
	ID     uuid.UUID
	UserID uuid.UUID

	// DeviceID is the app's own stable identifier for this installation.
	DeviceID string
	Platform Platform

	// PushToken is a credential for reaching this phone. It is loaded so the
	// push phase can use it, and is never put into any response. Nothing in
	// this package logs it (plans/phase8.md §8).
	PushToken string

	AppVersion string
	Active     bool

	LastSeenAt time.Time
	CreatedAt  time.Time
	UpdatedAt  time.Time
}

// ErrUnknownDevice is returned when a device id does not belong to the caller.
//
// It carries no information about whether the device exists for somebody else:
// the handler turns it into the same 404 an unknown id gets, so registrations
// cannot be probed (plans/phase8.md §§8, 40).
var ErrUnknownDevice = errors.New("notifications: no such device for this user")

// reminderNamespace seeds the deterministic reminder identifier below. A fixed
// random UUID, generated once; its only requirement is that it never changes,
// because changing it would rename every reminder already scheduled on every
// device and each would be scheduled a second time.
var reminderNamespace = uuid.MustParse("b7f6b7a2-3a5f-4f0e-9a1d-2c8f0a6d4e11")

// Reminder is one notification a device should schedule locally.
//
// It carries no wording. The title and body are composed on the device from
// packages/contracts/src/notification-labels.ts, the same way every other
// user-visible sentence in MeraCare is — one place where the phrasing lives,
// and no way for the server to accidentally put a medicine's name into
// something that appears on a lock screen (plans/phase8.md §§17, 47).
type Reminder struct {
	// ID is derived, not stored, and is the whole idempotency mechanism.
	ID uuid.UUID

	Type ReminderType

	SeniorID   uuid.UUID
	SeniorName string
	// SeniorTimezone is the IANA name the FireAt instant should be read in, so
	// the device shows "08:00" to a caregiver in another country too
	// (plans/phase8.md §32).
	SeniorTimezone string

	EntityType EntityType
	EntityID   uuid.UUID

	// DueAt is when the care itself is due.
	DueAt time.Time
	// FireAt is when the notification should appear: DueAt minus the type's
	// lead time.
	FireAt time.Time
}

// newReminder builds a reminder and derives its identifier.
//
// The identifier is a UUIDv5 over the recipient, the type, the subject, and the
// firing instant. That makes it a pure function of the plan: recomputing the
// same plan yields byte-identical ids, so a device that re-runs the whole
// reconciliation after every launch, every refresh, and every retry schedules
// each reminder exactly once (plans/phase8.md §§25, 26).
//
// FireAt is part of the identity on purpose. Moving a dose from 08:00 to 09:00
// must produce a different reminder, so the old one is cancelled and the new
// one scheduled, rather than a stale 07:45 alert surviving because its subject
// is unchanged (plans/phase8.md §22).
func newReminder(userID uuid.UUID, t ReminderType, senior Senior, entityID uuid.UUID, dueAt time.Time) Reminder {
	fireAt := dueAt.Add(-t.leadTime())

	seed := fmt.Sprintf("%s|%s|%s|%s", userID, t, entityID, fireAt.UTC().Format(time.RFC3339))

	return Reminder{
		ID:             uuid.NewSHA1(reminderNamespace, []byte(seed)),
		Type:           t,
		SeniorID:       senior.ID,
		SeniorName:     senior.DisplayName,
		SeniorTimezone: senior.Timezone,
		EntityType:     t.entityFor(),
		EntityID:       entityID,
		DueAt:          dueAt,
		FireAt:         fireAt,
	}
}
